File: backend/app/schemas/calendar.py
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, EmailStr, Field


class CalendarProvider(str, Enum):
    google = "google"
    apple = "apple"


class CalendarProviderInfo(BaseModel):
    id: CalendarProvider
    label: str
    auth_method: str
    scopes: List[str] = Field(default_factory=list)
    notes: Optional[str] = None


class CalendarAccountCreate(BaseModel):
    provider: CalendarProvider
    display_name: Optional[str] = None
    email: Optional[EmailStr] = None
    credentials: Dict[str, Any] = Field(
        default_factory=dict,
        description="Provider-specific auth payload (OAuth tokens or CalDAV password).",
    )
    config: Dict[str, Any] = Field(
        default_factory=dict,
        description="Additional configuration such as server URLs or selected calendars.",
    )


class CalendarAccount(BaseModel):
    account_id: str
    provider: CalendarProvider
    display_name: Optional[str] = None
    email: Optional[EmailStr] = None
    connected: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    last_sync_at: Optional[str] = None
    sync_cursor: Optional[str] = None


class CalendarAccountsResponse(BaseModel):
    accounts: List[CalendarAccount]


class ProvidersResponse(BaseModel):
    providers: List[CalendarProviderInfo]


# OAuth Flow Schemas
class OAuthStartResponse(BaseModel):
    auth_url: str
    state: str
    redirect_uri: str
    port: int


class OAuthCallbackRequest(BaseModel):
    code: str
    state: str


class OAuthCompleteResponse(BaseModel):
    account: CalendarAccount
    success: bool
    message: str


# Calendar Event Schema
class CalendarEvent(BaseModel):
    account_id: str
    event_id: str
    calendar_id: str
    title: str
    description: Optional[str] = None
    start_time: str
    end_time: str
    start_ts: int
    end_ts: int
    all_day: bool = False
    location: Optional[str] = None
    conference_url: Optional[str] = None
    visibility: Optional[str] = None
    status: str  # "confirmed", "tentative", "cancelled"
    organizer: Optional[Dict[str, Any]] = None
    attendees: Optional[List[Dict[str, Any]]] = None
    html_link: Optional[str] = None
    time_zone: Optional[str] = None
    created_at: str
    updated_at: str
    calendar_color: Optional[str] = None
    calendar_name: Optional[str] = None


class CalendarEventsResponse(BaseModel):
    events: List[CalendarEvent]


class CalendarListItem(BaseModel):
    account_id: str
    calendar_id: str
    provider: CalendarProvider
    name: str
    description: Optional[str] = None
    is_primary: bool = False
    access_role: Optional[str] = None
    background_color: Optional[str] = None
    foreground_color: Optional[str] = None
    time_zone: Optional[str] = None
    selected: bool = True
    created_at: str
    updated_at: str


class CalendarListResponse(BaseModel):
    calendars: List[CalendarListItem]


class CalendarSelectionUpdate(BaseModel):
    selected: bool

File: backend-bun/src/schemas/calendar.ts
import { z } from "zod";

// Calendar Provider
export type CalendarProvider = "google" | "apple";

// Request schemas
export const CalendarAccountCreateSchema = z.object({
  provider: z.union([z.literal("google"), z.literal("apple")]),
  display_name: z.string().optional(),
  email: z.string().optional(),
  credentials: z.record(z.string(), z.unknown()).optional(),
  config: z.record(z.string(), z.unknown()).optional(),
});

export type CalendarAccountCreateInput = z.infer<typeof CalendarAccountCreateSchema>;

export const CalendarSelectionUpdateSchema = z.object({
  selected: z.boolean(),
});

export type CalendarSelectionUpdateInput = z.infer<typeof CalendarSelectionUpdateSchema>;

export const OAuthCallbackRequestSchema = z.object({
  code: z.string(),
  state: z.string(),
});

export type OAuthCallbackRequestInput = z.infer<typeof OAuthCallbackRequestSchema>;

// Response types
export interface CalendarAccount {
  account_id: string;
  provider: CalendarProvider;
  display_name: string | null;
  email: string | null;
  connected: boolean;
  created_at: string;
  updated_at: string;
  last_sync_at: string | null;
}

export interface CalendarListItem {
  account_id: string;
  calendar_id: string;
  provider: string;
  name: string | null;
  description: string | null;
  is_primary: boolean;
  access_role: string | null;
  background_color: string | null;
  foreground_color: string | null;
  time_zone: string | null;
  selected: boolean;
}

export interface CalendarEvent {
  account_id: string;
  calendar_id: string;
  event_id: string;
  title: string | null;
  description: string | null;
  start_time: string | null;
  end_time: string | null;
  start_ts: number | null;
  end_ts: number | null;
  all_day: boolean;
  location: string | null;
  conference_url: string | null;
  visibility: string | null;
  status: string | null;
  organizer: Record<string, unknown> | null;
  attendees: Record<string, unknown>[] | null;
  html_link: string | null;
  time_zone: string | null;
  background_color: string | null;
  foreground_color: string | null;
}

export interface CalendarProviderInfo {
  id: CalendarProvider;
  label: string;
  auth_method: string;
  scopes: string[];
  notes: string;
}

export interface OAuthStartResponse {
  auth_url: string;
  state: string;
  redirect_uri: string;
  port: number;
}

export interface OAuthCompleteResponse {
  account: CalendarAccount | null;
  success: boolean;
  message: string;
}
